infer/environment.cc

```diff
@@ -1327,7 +1327,19 @@ core::TypePtr Environment::processBinding(core::Context ctx, const cfg::CFG &inW
                         }
                         break;
                     case cfg::CFG::MIN_LOOP_LET:
-                        if (!asGoodAs) {
+                        if (!hasType(ctx, bind.bind.variable)) {
+                            if (auto e =
+                                    ctx.beginError(bind.loc, core::errors::Infer::PinnedVariableAssignedBeforeLet)) {
+                                e.setHeader("Assignment to a variable declared via `{}` occurs before the `{}` "
+                                            "declaration has been reached",
```

Comment:
```suggestion
                                            "declaration",
```

non-blocking suggestion